### ISSUE TITLE
[Question] OPT 175B model served using alpa seems to have low benchmarking scores

### ISSUE THREAD

**Issue by zomux** [alpa-projects/alpa#559]
I'm serving the OPT 175B model using two GPU nodes with ray and alpa.

I tested three different small-scale tasks with different domains, the results seem to significantly underperform the openai davinci-002 model. (using temperature=0.1, top_p=1)

However, according to the OPT paper, the model is underperforming the openai one, but the performance is close.

Did you systematically benchmarked the opt 175B model on alpa on standard tasks and compare to the OPT paper?

Also, did you double check the completion is the same comparing to the result obtained using metaseq?

alpa commit: a559f811d858de025390bcb1a590865fe8761594

Example 1 (mathematical reasoning) 

OPT 175B
```
Question: Show 2 * x for x from 1 to 10
Answer: 2 4 6 8 10 12 14 16 18 20

Question: Show 11 * x for x from 1 to 3
Answer: 11 22 33

Question: Show 5 * x for x from 2 to 20
--- completion ---
Answer: 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 130 135 140 145 150 155 160 165 170 175 180 185 190 195 200 205 210 215 220 225 230 235 240 245 250 255 260 265 270 275 280 285 295 300 305 310 315 320 325
```

OpenAI Davinci-002
```
Question: Show 2 * x for x from 1 to 10
Answer: 2 4 6 8 10 12 14 16 18 20

Question: Show 11 * x for x from 1 to 3
Answer: 11 22 33

Question: Show 5 * x for x from 2 to 20
--- completion ---
Answer: 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 105 110 115 120 125 130 135 140 145 150 155 160 165 170 175 180 185 190 195 200
```

Example 2 (sentence compression)

OPT 175B
```
Text: Even though we were in competition for the grand prize, it seemed unfair.
Summarization: We were competing for the prize but it seemed unfair.

Text: The retreating guerrillas were soon pursued by the government forces.
Summarization: Government forces soon pursued the guerrillas.

Text: When a satellite passes through the belt, it is subject to bursts of gamma rays as the positrons annihilate electrons in its outer skin.
Summarization: A satellite passing though the belt is subject to bursts of gamma rays.

Text: The examiner who was kind let the student finish his lunch before disqualifying him.
--- completion ---
Summarization: An examiner let a student who had finished his exam finish lunch.
```

OpenAI Davinci-002
```
Text: Even though we were in competition for the grand prize, it seemed unfair.
Summarization: We were competing for the prize but it seemed unfair.

Text: The retreating guerrillas were soon pursued by the government forces.
Summarization: Government forces soon pursued the guerrillas.

Text: When a satellite passes through the belt, it is subject to bursts of gamma rays as the positrons annihilate electrons in its outer skin.
Summarization: A satellite passing though the belt is subject to bursts of gamma rays.

Text: The examiner who was kind let the student finish his lunch before disqualifying him.
--- completion ---
Summarization: The student was disqualified, but the examiner let him finish his lunch first.
```

**Comment by zhuohan123** [comment]
Hello,

Thanks for submitting the issue! We make sure that our implementation outputs exactly the same numerical results as [metaseq](https://github.com/facebookresearch/metaseq) at each time step. One potential reason for the results on standard tasks might be that we are using random sampling right now and in the OPT paper, they are using beam search. We are actively implementing the beam search version and it should achieve better performance. For now, can you try to sample multiple times and see whether there are correct results?

**Comment by zomux** [comment]
@zhuohan123 Oh thanks for the information. Let me check it with my benchmarking dataset.

**Comment by zhuohan123** [comment]
In addition, we tested the numerical correctness mainly with FP32. There might be some issues when running the model with FP16. Not sure whether it's possible or not, can you try to run the model with FP32 to see whether the results look better?

**Comment by zomux** [comment]
@zhuohan123 In my setting, I have 640GB GPU memory combined. Is it possible to run FP32?

**Comment by zhuohan123** [comment]
I see, in FP32 the model needs 175B * 4 = 700 GB of memory, so that is not enough. Let's try modifying the sampling methods first.

**Comment by merrymercy** [comment]
FP32 does not need 175GB * 4 = 700GB memory. The weights are stored as fp16 anyway. FP32 means we can do all computation in fp32.

I did some experiments on a small model: opt-2.7B.

1. First, I move all computation to fp32 by changing some types (e.g. https://github.com/alpa-projects/alpa/blob/da569d24e1aaed81b9196c93ac7af329630f8bbf/examples/opt_serving/model/opt_model.py#L397) to `jnp.float32`. But the generation results stay the same. This indicates the fp16/fp32 conversion may not be a problem.
2. Then, I compare the results generated by alpa and huggingface/pytorch with this script https://github.com/alpa-projects/alpa/blob/main/examples/opt_serving/benchmark/benchmark_text_gen.py 

By using the same sampling method (greedy) and random seed. They generate identical or very similar results. The first few tokens just after prompts are exactly the same. For some examples, they then generate slightly different tokens at the end of the sentences. This verifies alpa's numerical correctness. I guess the major difference is the sampling method and sampling parameters.

## Huggingface
The first 5-10 tokens are prompts.
```
ubuntu@ip-172-31-36-188:~/efs/alpa/examples/opt_serving/benchmark$ python3 benchmark_text_gen.py --model facebook/opt-2.7b --debug
input length: 8, output_length: 256, num_gpus: 1, speed: 34.58 tokens/s, tflops: 0.2062 tflops/s
['Computer science is the study of computation and information. It is the study of how information is processed and stored. It is the study of how computers work. It is the study of how information
is stored and retrieved. It is the study of how computers are used. It is the study of how computers are programmed. It is the study of how computers are controlled. It is the study of how computers are used. It is the study of how computers are programmed. It is the study of how computers are controlled. It is the study of how computers are used. It is the study of how computers are programmed. It is the study of how computers are controlled. It is the study of how computers are used. It is the study of how computers are programmed. It is the study of how computers are controlled. It is the study of how computers are used. It is the study of how computers are programmed. It is the study of how computers are controlled. It is the study of how computers are used. It is the study of
how computers are programmed. It is the study of how computers are controlled. It is the study of how computers are used. It is the study of how computers are programmed. It is the study of how computers are controlled. It is the']
input length: 13, output_length: 155, num_gpus: 1, speed: 33.88 tokens/s, tflops: 0.2020 tflops/s
["Ion Stoica is a Romanian-American computer scientist specializing in computer security. He is the founder and CEO of the company, which he founded in 2012.\n\nCareer\nStoica is a computer scientist and a computer security expert. He is the founder and CEO of the company, which he founded in 2012.\n\nStoica is a professor at the University of California, Berkeley, where he teaches courses on
computer security and cryptography. He is also a visiting professor at the University of California, Santa Barbara.\n\nStoica is a member of the National Academy of Engineering.\n\nAwards and honors\nStoica was awarded the IEEE Computer Society's 2016 IEEE Computer Society Medal for his contributions to computer security.\n\nReferences"]
input length: 9, output_length: 256, num_gpus: 1, speed: 33.75 tokens/s, tflops: 0.2012 tflops/s
['The University of California, Berkeley is a public research university located in Berkeley, California, United States. It is the largest university in the state of California and the second-largest in the United States by enrollment. The university was founded in 1868 as the California Agricultural College, and was renamed the University of California in 1873. The university is a member of the Association of American Universities and the Association of American Universities.\n\nThe university is home to the California Institute of Technology, the oldest and largest public research university in the United States. The university is also home to the California School of Professional Psychology, the California School of Public Health, the California School of Social Work, the California School of Dentistry, the California School of Medicine, the California School of Optometry, the California School of Nursing, the California School of Pharmacy, the California School of Veterinary Medicine, the California School of Law, the California School of the Arts, the California School of the Arts, the California School of the Environment, the California School of Engineering, the
California School of Law, the California School of Medicine, the California School of Nursing, the California School of Optometry, the California School of Pharmacy, the California School of Veterinary Medicine, the California School of Law, the California']
input length: 9, output_length: 256, num_gpus: 1, speed: 33.71 tokens/s, tflops: 0.2010 tflops/s
['Today is a good day and I want to thank you all for your support.\nI am so grateful for all of you.\nI am so grateful for the love and support of my family and friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so']
input length: 9, output_length: 256, num_gpus: 1, speed: 33.61 tokens/s, tflops: 0.2004 tflops/s
['What is the valuation of Databricks?\n\nDatabricks is a company that is valued at $1.5 billion. The company was founded in 2013 and is based in Mountain View, California.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1']
Model: facebook/opt-2.7b  Device: cuda  Dummy: False  Load (s): 107.44  Autoregressive: True  Batchsize: 1  #Microbatches: 1  #Stages: 2  Decoder step length: 1  TFlops: 0.1011  Compute TFlops: 0.1011  Speed (token/s): 16.95  latency (32 token): 1.89
```

## Alpa
```
ubuntu@ip-172-31-36-188:~/efs/alpa/examples/opt_serving/benchmark$ python3 benchmark_text_gen.py --model alpa/opt-2.7b --debug
Load model alpa/opt-2.7b ... (This can take several minutes for very large models)
2022-07-01 14:54:22,601 INFO worker.py:840 -- Connecting to existing Ray cluster at address: 172.31.36.188:6379
input length: 8, output_length: 256, num_gpus: 8, speed: 50.03 tokens/s, tflops: 0.0373 tflops/s
['Computer science is the study of computation and information. It is the study of how information is processed and stored. It is the study of how computers work. It is the study of how computers are used. It is the study of how computers are programmed. It is the study of how computers are used. It is the study of how computers are programmed. It is the study of how computers are used. It is
the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of how computers are used. It is the study of']
input length: 13, output_length: 155, num_gpus: 8, speed: 49.58 tokens/s, tflops: 0.0369 tflops/s
["Ion Stoica is a Romanian-American computer scientist specializing in computer security. He is the founder and CEO of the company, which he founded in 2012.\n\nCareer\nStoica is a computer scientist and a computer security expert. He is the founder and CEO of the company, which he founded in 2012.\n\nStoica is a professor at the University of California, Berkeley, where he teaches courses on
computer security and cryptography. He is also a visiting professor at the University of California, Santa Barbara.\n\nStoica is a member of the National Academy of Engineering.\n\nAwards and honors\nStoica was awarded the IEEE Computer Society's 2016 IEEE Computer Society Medal for his contributions to computer security.\n\nReferences"]
input length: 9, output_length: 256, num_gpus: 8, speed: 49.45 tokens/s, tflops: 0.0369 tflops/s
['The University of California, Berkeley is a public research university located in Berkeley, California, United States. It is the largest university in the state of California and the second-largest in the United States by enrollment. The university was founded in 1868 as the California Agricultural College, and was renamed the University of California in 1873. The university is a member of the Association of American Universities and the Association of American Universities.\n\nThe university is home to the California Institute of Technology, the oldest and largest public research university in the United States. The university is also home to the National Center for Supercomputing Applications, the National Center for Atmospheric Research, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Livermore National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the
Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence Berkeley National Laboratory, the Lawrence
Berkeley National Laboratory, the Lawrence Berkeley National']
input length: 9, output_length: 256, num_gpus: 8, speed: 49.94 tokens/s, tflops: 0.0372 tflops/s
['Today is a good day and I want to thank you all for your support.\nI am so grateful for all of you.\nI am so grateful for the love and support of my family and friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so grateful for the love and support of my friends.\nI am so grateful for the love and support of my fans.\nI am so']
input length: 9, output_length: 256, num_gpus: 8, speed: 50.23 tokens/s, tflops: 0.0374 tflops/s
['What is the valuation of Databricks?\n\nDatabricks is a company that is valued at $1.5 billion. The company was founded in 2012 and is based in Mountain View, California.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1.2 billion in funding and has a valuation of $1.5 billion.\n\nThe company has raised $1']
Model: alpa/opt-2.7b  Device: cuda  Dummy: False  Load (s): 35.72  Autoregressive: True  Batchsize: 1  #Microbatches: 1  #Stages: 2  Decoder step length: 1  TFlops: 0.0186  Compute TFlops: 0.0390  Speed (token/s): 24.92  latency (32 token): 1.28
```

**Comment by merrymercy** [comment]
Beam search is supported https://github.com/alpa-projects/alpa/blob/724a148cf45b3800a10788d8e27b5fac927bb66b/examples/opt_serving/benchmark/benchmark_text_gen.py#L56

**Comment by cokuehuang** [comment]
@merrymercy  Hi, is this speed ok ? Ray cluster has two nodes, 4 * T4 16G GPU and 128G memory per node.
`(alpa) root@ubuntu-Default-string:/home/alpa/examples/opt_serving/benchmark# python benchmark_text_gen.py --model alpa/opt-30b --path /disk1/ --debug
Load model alpa/opt-30b ... (This can take several minutes for very large models)
input length: 8, output_length: 256, num_gpus: 8, speed: 4.36 tokens/s, tflops: 0.0342 tflops/s
['Computer science is the study of computation and its applications. Computers are used to solve problems that are too large or too complex for humans to solve. The field of computer science is concerned with the theoretical foundations of computation, the design of algorithms, and the construction of computers and computer systems.\n\nComputer science is a vast field that includes the study of algorithms, data structures, computer architecture, operating systems, programming languages, and computer networks. Computer science is a multidisciplinary field that involves the application of mathematics, logic, and engineering to the design and analysis of computer systems.\n\nComputer science is a vast field that includes the study of algorithms, data structures, computer architecture, operating systems, programming languages, and computer networks. Computer science is a multidisciplinary field that involves the application of mathematics, logic, and engineering to the design and analysis of computer systems.\n\nComputer science is the study of computation and its applications. Computers are used to solve problems that are too large or too complex for humans to solve. The field of computer science is concerned with the theoretical foundations of computation, the design of algorithms, and the construction of computers and computer systems.\n\nComputer science is the study of computation and its applications. Computers are used to solve problems']`

**Comment by merrymercy** [comment]
@cokuehuang The speed looks okay. There are still room for improvement. We are actively working on improving the latency/throughout.

**Comment by merrymercy** [comment]
closed due to inactivity and the added support of beam search
